--- a/Core/HLE/sceKernelInterrupt.cpp	1970-01-01 09:00:01.000000000 +0900
+++ b/Core/HLE/sceKernelInterrupt.cpp	2026-08-18 16:16:12.877481539 +0900
@@ -996,6 +996,58 @@
 	return hleLogVerbose(Log::sceKernel, toupper(c));
 }

+static u32 sysclib_tolower(u32 c) {
+	return hleLogVerbose(Log::sceKernel, tolower(c));
+}
+
+static u32 sysclib_strtoul(u32 strPtr, u32 endPtrPtr, int base) {
+	if (!Memory::IsValidNullTerminatedString(strPtr)) {
+		return hleLogError(Log::sceKernel, 0, "invalid address");
+	}
+	const char* str = Memory::GetCharPointer(strPtr);
+	char* end = nullptr;
+	u32 result = (u32)strtoul(str, &end, base);
+	if (Memory::IsValidRange(endPtrPtr, 4))
+		Memory::WriteUnchecked_U32(strPtr + (u32)(end - str), endPtrPtr);
+	return hleLogVerbose(Log::sceKernel, result);
+}
+
+static u32 sysclib_memchr(u32 s, int c, u32 size) {
+	if (Memory::IsValidRange(s, size)) {
+		const void *p = memchr(Memory::GetPointerUnchecked(s), c, size);
+		if (p == nullptr)
+			return hleLogVerbose(Log::sceKernel, 0);
+		return hleLogVerbose(Log::sceKernel, s + (u32)((const u8 *)p - Memory::GetPointerUnchecked(s)));
+	}
+	return hleLogError(Log::sceKernel, 0, "invalid address");
+}
+
+static u32 sysclib_strncat(u32 dest, u32 src, u32 size) {
+	if (!Memory::IsValidAddress(dest) || !Memory::IsValidNullTerminatedString(src)) {
+		return hleLogError(Log::sceKernel, 0, "invalid address");
+	}
+
+	u32 destLen = 0;
+	u32 destSize = Memory::ClampValidSizeAt(dest, 0x7FFFFFFF);
+	const u8 *destScan = Memory::GetPointerUnchecked(dest);
+	while (destLen < destSize && destScan[destLen] != 0)
+		destLen++;
+
+	const u8 *srcp = Memory::GetPointerUnchecked(src);
+	u8 *destp = Memory::GetPointerWriteUnchecked(dest) + destLen;
+	u32 srcSize = Memory::ClampValidSizeAt(src, size);
+	u32 i;
+	for (i = 0; i < srcSize && i < size; ++i) {
+		u8 c = srcp[i];
+		if (c == 0)
+			break;
+		*destp++ = c;
+	}
+	*destp = 0;
+
+	return hleLogVerbose(Log::sceKernel, dest);
+}
+
 const HLEFunction SysclibForKernel[] =
 {
 	{0xAB7592FF, &WrapU_UUU<sysclib_memcpy>,                   "memcpy",                              'x', "xxx",    HLE_KERNEL_SYSCALL },
@@ -1015,6 +1067,10 @@
 	{0x4C0E0274, &WrapU_UI<sysclib_strrchr>,                   "strrchr",                             'x', "xx",    HLE_KERNEL_SYSCALL },
 	{0xCE2F7487, &WrapU_U<sysclib_toupper>,                    "toupper",                             'x', "x",     HLE_KERNEL_SYSCALL },
 	{0XC2145E80, &WrapI_UIU<sysclib_snprintf>,                 "snprintf",                            'i', "xx",     HLE_KERNEL_SYSCALL },
+	{0x3EC5BBF6, &WrapU_U<sysclib_tolower>,                    "tolower",                             'x', "x",     HLE_KERNEL_SYSCALL },
+	{0x6A7900E1, &WrapU_UUI<sysclib_strtoul>,                  "strtoul",                             'x', "xxi",    HLE_KERNEL_SYSCALL },
+	{0x68A78817, &WrapU_UIU<sysclib_memchr>,                   "memchr",                              'x', "xii",    HLE_KERNEL_SYSCALL },
+	{0xD3D1A3B9, &WrapU_UUU<sysclib_strncat>,                  "strncat",                             'x', "xxi",    HLE_KERNEL_SYSCALL },
 };

 void Register_Kernel_Library()
